compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PolymorphicSignatureLowering.kt

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.ir.copyTypeParametersFrom
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.JvmLoweredDeclarationOrigin
+import org.jetbrains.kotlin.config.LanguageFeature
+import org.jetbrains.kotlin.ir.builders.declarations.addValueParameter
+import org.jetbrains.kotlin.ir.builders.declarations.buildFun
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrCallImpl
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.isNullableAny
+import org.jetbrains.kotlin.ir.util.dump
+import org.jetbrains.kotlin.ir.util.hasAnnotation
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.name.FqName
+
+internal val polymorphicSignaturePhase = makeIrFilePhase(
+    ::PolymorphicSignatureLowering,
+    name = "PolymorphicSignature",
+    description = "Replace polymorphic methods with fake ones according to types at the call site"
+)
+
+private val polymorphicSignatureAnnotation = FqName("java.lang.invoke.MethodHandle.PolymorphicSignature")
+
+class PolymorphicSignatureLowering(val context: JvmBackendContext) : IrElementTransformerVoid(), FileLoweringPass {
+    override fun lower(irFile: IrFile) {
+        if (context.state.languageVersionSettings.supportsFeature(LanguageFeature.PolymorphicSignature))
+            irFile.transformChildrenVoid()
+    }
+
+    // If the return type is Any?, then it is also polymorphic (e.g. MethodHandle.invokeExact
+    // has polymorphic return type, while VarHandle.compareAndSet does not).
+    override fun visitTypeOperator(expression: IrTypeOperatorCall): IrExpression =
+        (expression.argument as? IrCall)?.takeIf { it.type.isNullableAny() }?.transform(expression.typeOperand)
```

Comment:
Do we need to check `expression.operator` here for `as`/`as?`? Otherwise looks like `invokeExact(...) is X` would cause the return type to be incorrectly specialized to `X`.

Reply:
Yes, looks like we do.